File: 2021/day14/day14.py
from collections import defaultdict


def read_input():
    with open("input.in", "r") as file:
        line = file.readline().strip()
        file.readline()
        rules = {rule[0]: rule[1] for rule in [line.strip().split(" -> ") for line in file.readlines()]}
    template = defaultdict(int)
    for i in range(len(line) - 1):
        template[line[i:i+2]] += 1
    return template, rules


def insert_once(template, rules):
    pairs = [(key, value) for key, value in template.items() if value != 0]
    for pair, value in pairs:
        template[pair] -= value
        element = rules[pair]
        template[pair[0] + element] += value
        template[element + pair[1]] += value


def insert_n(template, rules, n):
    for i in range(n):
        insert_once(template, rules)


def get_most_least_common_element(template):
    element_counts = defaultdict(int)
    for pair, value in template.items():
        element_counts[pair[0]] += value
        element_counts[pair[1]] += value
    for element, count in element_counts.items():
        element_counts[element] = -(count // -2)  # Ceiling division by 2
    counts = element_counts.values()
    return max(counts), min(counts)


def part1():
    template, rules = read_input()
    insert_n(template, rules, 10)
    most_common_count, least_common_count = get_most_least_common_element(template)
    print(f"Part 1: {most_common_count - least_common_count}")


def part2():
    template, rules = read_input()
    insert_n(template, rules, 40)
    most_common_count, least_common_count = get_most_least_common_element(template)
    print(f"Part 2: {most_common_count - least_common_count}")


if __name__ == "__main__":
    part1()
    part2()

File: 2021/day14/day14.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <sstream>
#include <fstream>
#include <unordered_map>

using namespace std;

typedef long long ll;

void part1(const pair<unordered_map<string, ll>, unordered_map<string, string>>& polymerAndRules);
void part2(const pair<unordered_map<string, ll>, unordered_map<string, string>>& polymerAndRules);
pair<unordered_map<string, ll>, unordered_map<string, string>> readInput();
void insertOnce(unordered_map<string, ll> &polymer, const unordered_map<string, string> &rules);
void insertN(unordered_map<string, ll> &polymer, const unordered_map<string, string> &rules, int n);
pair<ll, ll> getMostLeastCommonElement(const unordered_map<string, ll> &polymer);


int main() {
    pair<unordered_map<string, ll>, unordered_map<string, string>> polymerAndRules = readInput();
    part1(polymerAndRules);
    part2(polymerAndRules);
    return 0;
}

void part1(const pair<unordered_map<string, ll>, unordered_map<string, string>>& polymerAndRules) {
    unordered_map<string, ll> polymer = polymerAndRules.first;
    unordered_map<string, string> rules = polymerAndRules.second;
    insertN(polymer, rules, 10);
    pair<ll, ll> counts = getMostLeastCommonElement(polymer);
    ll answer = counts.first - counts.second;
    cout << "Part 1: " << answer << endl;
}

void part2(const pair<unordered_map<string, ll>, unordered_map<string, string>>& polymerAndRules) {
    unordered_map<string, ll> polymer = polymerAndRules.first;
    unordered_map<string, string> rules = polymerAndRules.second;
    insertN(polymer, rules, 40);
    pair<ll, ll> counts = getMostLeastCommonElement(polymer);
    ll answer = counts.first - counts.second;
    cout << "Part 1: " << answer << endl;
}

pair<unordered_map<string, ll>, unordered_map<string, string>> readInput() {
    unordered_map<string, ll> polymer;
    unordered_map<string, string> rules;
    string line;

    // Process first line
    cin >> line;
    for (int i = 0; i < line.size() - 1; i++) {
        string pairOfElements = line.substr(i, 2);
        if (polymer.find(pairOfElements) == polymer.end()) {
            polymer[pairOfElements] = 0;
        }
        polymer[pairOfElements]++;
    }

    // Process rules
    string pairOfElements, value;
    while (cin >> pairOfElements) {
        cin >> line; cin >> value;
        rules[pairOfElements] = value;
    }

    return {polymer, rules};
}

void insertOnce(unordered_map<string, ll> &polymer, const unordered_map<string, string> &rules) {
    vector<pair<string, ll>> temp;
    for (const auto& pairOfElementsCount : polymer) {
        if (pairOfElementsCount.second == 0) {
            continue;
        }
        temp.emplace_back(pairOfElementsCount);
    }
    for (const auto& pairOfElementsCount : temp) {
        string pairOfElements = pairOfElementsCount.first;
        ll count = pairOfElementsCount.second;
        polymer[pairOfElements] -= count;
        const string& element = rules.at(pairOfElements);
        string firstPair = pairOfElements[0] + element, secondPair = element + pairOfElements[1];
        if (polymer.find(firstPair) == polymer.end()) {
            polymer[firstPair] = 0;
        }
        polymer[firstPair] += count;
        if (polymer.find(secondPair) == polymer.end()) {
            polymer[secondPair] = 0;
        }
        polymer[secondPair] += count;
    }
}

void insertN(unordered_map<string, ll> &polymer, const unordered_map<string, string> &rules, int n) {
    for (int i = 0; i < n; i++) {
        insertOnce(polymer, rules);
    }
}

pair<ll, ll> getMostLeastCommonElement(const unordered_map<string, ll> &polymer) {
    unordered_map<char, ll> elementCount;
    for (const auto& pairOfElementsCount : polymer) {
        string pairOfElements = pairOfElementsCount.first;
        ll count = pairOfElementsCount.second;
        for (int i = 0; i < 2; i++) {
            char element = pairOfElements[i];
            if (elementCount.find(element) == elementCount.end()) {
                elementCount[element] = 0;
            }
            elementCount[element] += count;
        }
    }
    vector<ll> counts;
    for (auto elementAndCount : elementCount) {
        ll count = elementAndCount.second;
        counts.push_back(1 + ((count - 1) / 2));  // Ceiling division, avoids integer overflow too
    }
    ll maxCount = *max_element(counts.begin(), counts.end());
    ll minCount = *min_element(counts.begin(), counts.end());
    return {maxCount, minCount};
}
